virtool/uploads/api.py

```diff
@@ -53,59 +58,137 @@ async def naive_writer(req, file_id):
             size += len(chunk)
             await handle.write(chunk)
 
-    return size
+    return size, virtool.utils.timestamp()
 
 
-@routes.post("/upload/{file_type}", permission="upload_file")
+@routes.post("/api/uploads", permission="upload_file")
 async def upload(req):
-    db = req.app["db"]
-
-    file_type = req.match_info["file_type"]
-
-    if file_type not in FILE_TYPES:
-        return not_found()
+    db = req.app["postgres"]
+    upload_type = req.query.get("type")
 
     errors = naive_validator(req)
 
     if errors:
         return invalid_query(errors)
 
-    filename = req.query["name"]
+    name = req.query["name"]
+
+    if upload_type not in UPLOAD_TYPES:
+        return bad_request("Unsupported upload type")
+
+    upload_ = await virtool.uploads.db.create(db, name, upload_type, user=req["client"].user_id)
+
+    upload_id = upload_["id"]
 
-    document = await virtool.files.db.create(
-        db,
-        filename,
-        file_type,
-        user_id=req["client"].user_id
-    )
+    async with AsyncSession(db) as session:
+        try:
+            result = (await session.execute(select(Upload).filter_by(id=upload_id))).scalar()
+            size, uploaded_at = await naive_writer(req, upload_id, name)
 
-    file_id = document["id"]
+            result.size = size
+            result.name_on_disk = f"{upload_id}-{name}"
+            result.uploaded_at = uploaded_at
 
-    try:
-        size = await naive_writer(req, file_id)
+            await session.commit()
 
-        await db.files.update_one({"_id": file_id}, {
-            "$set": {
+            upload_.update({
                 "size": size,
-                "ready": True
+                "name_on_disk": f"{upload_id}-{name}",
+                "uploaded_at": uploaded_at
+            })
+
+            logger.debug(f"Upload succeeded: {upload_id}")
+
+            headers = {
+                "Location": f"/api/uploads/{upload_id}"
             }
-        })
 
-        logger.debug(f"Upload succeeded: {file_id}")
+            return json_response(upload_, status=201, headers=headers)
+        except asyncio.CancelledError:
+            logger.debug(f"Upload aborted: {upload_id}")
+
+            await session.delete(result)
+            await session.commit()
+
+            return aiohttp.web.Response(status=499)
+
+
+@routes.get("/api/uploads")
+async def find(req):
+    db = req.app["postgres"]
+    upload_ = list()
+    filters = list()
+    user = req.query.get("user")
+    type_ = req.query.get("type")
```

Comment:
You name this `upload_type` in other conceptually similar situations. Make this consistent. If you are having to use a trailing `_` because your name shadows a built-in (`type`), you probably need a more informative name.